marlin-dlog/protocol/src/prover.rs

```diff
@@ -0,0 +1,483 @@
+        argument.commit_points(&[w_comm, za_comm, zb_comm])?;
+
+        // sample alpha, eta oracles
+        oracles.alpha = argument.challenge();
+        oracles.eta_a = argument.challenge();
+        oracles.eta_b = argument.challenge();
+        oracles.eta_c = argument.challenge();
+
+        let mut apow = E::Fr::one();
+        let mut r: Vec<E::Fr> = (0..index.h_group.size()).map
+        (
+            |i|
+            {
+                if i > 0 {apow *= &oracles.alpha}
+                apow
+            }
+        ).collect();
+        r.reverse();
+        let ra = DensePolynomial::<E::Fr>::from_coefficients_vec(r);
+
+        // compute first sumcheck argument polynomials
+        // --------------------------------------------------------------------
+
+        let (h1, mut g1) = Self::sumcheck_1_compute (index, &ra, &zv, &z, &oracles)?;
+        if !g1.coeffs[0].is_zero() {return Err(ProofError::SumCheck)}
+        g1.coeffs.remove(0);
+
+        // commit to H1 & G1 polynomials and
+        let h1_comm = index.srs.commit(&h1, index.h_group.size()*2-2)?;
+        let g1_comm = index.srs.commit(&g1, index.h_group.size()-1)?;
+
+        // absorb H1, G1 polycommitments
+        argument.commit_points(&[h1_comm, g1_comm])?;
+        // sample beta[0] oracle
+        oracles.beta[0] = argument.challenge();
+
+        // compute second sumcheck argument polynomials
+        // --------------------------------------------------------------------
+
+        let (h2, mut g2) = Self::sumcheck_2_compute (index, &ra, &oracles)?;
+        let sigma2 = g2.coeffs[0];
+        g2.coeffs.remove(0);
+
+        // absorb sigma2 scalar
+        argument.commit_scalars(&[sigma2]);
+        // sample beta[1] oracle
+        oracles.beta[1] = argument.challenge();
+
+        // compute third sumcheck argument polynomials
+        // --------------------------------------------------------------------
+
+        let (h3, mut g3) = Self::sumcheck_3_compute (index, &oracles)?;
+        let sigma3 = g3.coeffs[0];
+        g3.coeffs.remove(0);
+
+        // absorb sigma3 scalar
+        argument.commit_scalars(&[sigma3]);
+        // sample beta[2] & batch oracles
+        oracles.beta[2] = argument.challenge();
+        oracles.batch = argument.challenge();
+
+        // construct the proof
+        // --------------------------------------------------------------------
+
+        Ok(ProverProof
+        {
+            // polynomial commitments
+            w_comm,
+            za_comm,
+            zb_comm,
+            h1_comm,
+            g1_comm,
+            h2_comm: index.srs.commit(&h2, index.h_group.size()-1)?,
+            g2_comm: index.srs.commit(&g2, index.h_group.size()-1)?,
+            h3_comm: index.srs.commit(&h3, index.k_group.size()*6-6)?,
+            g3_comm: index.srs.commit(&g3, index.k_group.size()-1)?,
+
+            // polynomial commitment batched opening proofs
+            proof1: index.srs.open
```

Comment:
All the opening proofs should be batched into one.

The way to do this is described [here in the first section](https://www.overleaf.com/project/5ddbf8f6b95ef30001b7a31d). The idea is you take all the polynomials (w, za, zb, h1, g1, g2, h2, g3, h3, matrix polynomials) and sum them with the Horner linear combination like F := sum_i xi^i * f_i. You also pick a random r_j for each evaluation point (the betas).

Then you use the inner product argument on < F_i, sum_j r_j pows(beta_j) > where pows(x) = (1, x, x^2, ...).

Reply:
The following has been committed into this branch:

1. D-log polycommitment with degree bouns with shifted and unshifted commitments as per Marlin
2. Batching d-log polycommitment openings across multiple evaluation points
3. D-log polycommitment tests
4. Protocol with d-log polycommitment unit tests